engine/common/version/version_control.go

```diff
@@ -0,0 +1,314 @@
+	}
+
+	startHeight := v.startHeight.Load()
+	// Check if the start height is set and the height is less than the start height. If so, return false indicating that the height is not compatible.
+	if startHeight != NoHeight && height < startHeight {
+		return false, nil
+	}
+
+	endHeight := v.endHeight.Load()
+	// Check if the end height is set and the height is greater than the end height. If so, return false indicating that the height is not compatible.
+	if endHeight != NoHeight && height > endHeight {
+		return false, nil
+	}
+
+	// If none of the above conditions are met, the height is compatible.
+	return true, nil
+}
+
+// AddVersionUpdatesConsumer adds a consumer for version update events.
+func (v *VersionControl) AddVersionUpdatesConsumer(consumer VersionControlConsumer) {
+	v.Lock()
+	defer v.Unlock()
+
+	v.consumers = append(v.consumers, consumer)
+}
+
+// processEvents is a worker that processes block finalized events.
+func (v *VersionControl) processEvents(
+	ctx irrecoverable.SignalerContext,
+	ready component.ReadyFunc,
+) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-v.finalizedHeightNotifier.Channel():
+			v.blockFinalized(ctx, v.finalizedHeight.Value())
+		}
+	}
+}
+
+// blockFinalized processes a block finalized event and updates the version control state.
+func (v *VersionControl) blockFinalized(
+	ctx irrecoverable.SignalerContext,
+	newFinalizedHeight uint64,
+) {
+	lastProcessedHeight := v.lastProcessedHeight.Load()
+	if lastProcessedHeight >= newFinalizedHeight {
+		// already processed this or a higher version beacon
+		return
+	}
+
+	for height := lastProcessedHeight + 1; height <= newFinalizedHeight; height++ {
+		vb, err := v.versionBeacons.Highest(height)
+		if err != nil {
+			v.log.Err(err).
+				Uint64("height", height).
+				Msg("Failed to get highest version beacon for version control")
+
+			ctx.Throw(
+				fmt.Errorf(
+					"failed to get highest version beacon for version control: %w",
+					err))
+			return
+		}
+
+		if vb == nil {
+			// no version beacon found
+			// this is unexpected as there should always be at least the
+			// starting version beacon, but not fatal.
+			// It can happen if the node starts before bootstrap is finished.
+			// TODO: remove when we can guarantee that there will always be a version beacon
+			v.log.Info().
+				Uint64("height", height).
+				Msg("No version beacon found for version control")
+			continue
+		}
+
+		v.lastProcessedHeight.Store(height)
+
+		// version boundaries are sorted by version
+		for _, boundary := range vb.VersionBoundaries {
+			ver, err := boundary.Semver()
+			if err != nil || ver == nil {
+				// this should never happen as we already validated the version beacon
+				// when indexing it
+				ctx.Throw(
+					fmt.Errorf(
+						"failed to parse semver: %w",
+						err))
+				return
+			}
+
```

Comment:
Thinking about this some more, the simplest thing to do is:
1. If `v.endHeight` is set, check if `height > v.endHeight`. If it is, we don't need to continue since it's outside of our compatible range.
2. Use a local variable `endHeight = NoHeight` while searching boundaries in `vb`. Then set `v.endHeight = endHeight` after the loop. That way deletions and updates are handled automatically. This is effectively how the `StopControl` works: https://github.com/onflow/flow-go/blob/ac44d356ec285ad79fdcd52bd1ff4f369bf4b61a/engine/execution/ingestion/stop/stop_control.go#L635

Reply:
We'll also need to come up with a way to communicate the deletions to consumers